src/algebra/jordan/basic.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2021 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.ring.basic
+import algebra.lie.of_associative
+import data.real.basic
+import linear_algebra.basic
+
+/-!
+# Jordan algebras
+
+We define a set of conditions, formulated in terms of commuting multiplication operators, for the
+multiplication on a non-unital, non-associative semiring to be a Jordan multiplication. When the
+multiplication is commutative, these take a particularly simple form.
+
+A real Jordan algebra `A` can be introduced by
+```
+variables {A : Type*} [non_unital_non_assoc_ring A] [module ℝ A] [smul_comm_class ℝ A A]
+  [is_scalar_tower ℝ A A] [comm_jordan A]
+```
+
+## Main results
+
+- lin_jordan : Liniarisation of the commutative Jordan axiom
+
+## Implementation notes
+
+We shall primarily be interested in linear Jordan algebras (i.e. over rings of characteristic not
+two) leaving quadratic algebras to those better versed in that theory.
+
+The conventional way to linearise the Jordan axiom is to equate coefficients (more formally, assume
+that the axiom holds in all field extensions). For simplicity we use brute force algebraic expansion
+and substitution instead.
+
+## References
+
+* [Hanche-Olsen and Størmer, Jordan Operator Algebras][hancheolsenstormer1984]
+* [Cabrera García and Rodríguez Palacios, Non-associative normed algebras. Volume 1]
+  [cabreragarciarodriguezpalacios2014]
+-/
+
+set_option old_structure_cmd true
+
+/-- A not-necessarily-unital, not-necessarily-associative ring. -/
+@[protect_proj, ancestor add_comm_group non_unital_non_assoc_semiring ]
+class non_unital_non_assoc_ring (α : Type*) extends
+  add_comm_group α, non_unital_non_assoc_semiring α
```

Comment:
This belongs in algebra/ring/basic, along with the other two typeclasses

Reply:
I've moved `non_unital_non_assoc_ring` into `algebra.ring.basic` but now `leanpkg build` is repeatedly failing with:
```
Run leanpkg configure
configuring mathlib 0.1
Error: /home/lean/actions-runner/_work/mathlib/mathlib/src/number_theory/quadratic_reciprocity.lean:236:5:
(deterministic) timeout
Error: Process completed with exit code 1.
```
It's not immediately clear to me how this is related.

Reply:
I recommend adding these classes in a standalone PR, and waiting for github to build them rather than attempting to build them yourself.

Reply:
Thanks. I've added `non_unital_non_assoc_ring` here: #11124 but it has the same [problem](https://github.com/leanprover-community/mathlib/runs/4656806136?check_suite_focus=true)

Reply:
I've managed to get it to build now.